examples/experimental/imagenet_nas_arch_pytorch/data.py

```diff
@@ -0,0 +1,118 @@
+import logging
+import os
+import shutil
+
+import numpy as np
+from torch.utils.data import Dataset
+
+from torchvision import datasets, transforms
+from determined.util import download_gcs_blob_with_backoff
+from io import BytesIO
+from PIL import Image
+from google.cloud import storage
+
+
+# Helper function from https://pytorch.org/docs/stable/_modules/torchvision/datasets/folder.html#ImageFolder
+def load_image(path):
+    with open(path, 'rb') as f:
+        img = Image.open(f)
+        return img.convert('RGB')
+
+# Helper functions for GCP from https://cloud.google.com/storage/docs/listing-objects#code-samples
+def list_blobs(storage_client, bucket_name, prefix=None):
+    """Lists all the blobs in the bucket."""
+    blobs = storage_client.list_blobs(bucket_name, prefix=prefix)
+    paths = []
+    for b in blobs:
+        paths.append(b.name)
+    return paths
+
+def create_file_dirs(target_path):
+    destination_dir = target_path[0:target_path.rfind('/')]
+    if not os.path.exists(destination_dir):
+        try:
+            os.makedirs(destination_dir)
+        except:
+            assert os.path.exists(destination_dir)
+            pass
+
+def download_entire_imagenet(data_dir='/mnt/data'):
+    client = storage.Client()
+    for split in ['validation', 'train']:
+        prefix = os.path.join('imagenet/imagenet', split)
+        blobs = client.list_blobs(client, 'determined-ai-datasets', prefix=prefix)
+        for b in blobs:
+            blob.download_to_filename(os.path.join(data_dir, b.name))
+
+class ImageNetDataset(Dataset):
+    def __init__(self, split, bucket_name, transform=None):
+        """
+        Args:
+            split: train or validation split to return the right dataset
+            directory: root directory for imagenet where "train" and "validation" folders reside
+        """
+        assert split in ["train", "validation"], "split {} not in (train, validation)".format(split)
+        self._split = split
+        self._bucket_name = bucket_name
+        self._target_dir = '/mnt/data'
+        self._source_dir = os.path.join('imagenet/imagenet', self._split)
+        self._transform = transform
+        if self._bucket_name is not None:
+            self._storage_client = storage.Client()
+            self._bucket = self._storage_client.bucket(bucket_name)
+            # When the dataset is first initialized, we'll loop through to catalogue the classes (subdirectories)
+            # This step might take a long time.
+            self._imgs_paths = list_blobs(self._storage_client, self._bucket_name, prefix=self._source_dir)
+
+            self._labels = []
+            self._subdir_to_class = {}
+            class_count = 0
+            for path in self._imgs_paths:
```

Comment:
when we loop through to get the path "names" can we also loop through to get some of these counts to prevent looping through 1mil + images?

Reply:
good suggestion, i combined into one loop